### PR TITLE
Add MustHasReducers trait.

### PR DESCRIPTION
**Pull request by reconditesea**
For #1248.

### CODE REVIEW

File: scalding-core/src/main/scala/com/twitter/scalding/typed/WithReducers.scala [hunk 1]
```diff
@@ -24,6 +24,18 @@ trait HasReducers {
 }
 
 /**
+ * used for types that must know how many reducers they need
+ * e.g. Sketched
+ */
+trait MustHasReducers {
+  def reducers: Option[Int]
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/WithReducers.scala]: Let's not do the runtime exception. Instead can we do this:

```scala
trait MustHaveReducers extends HasReducers {
  def reducers: Some[Int]
}
```
This should show that reducers is a Some[T] which is a subclass of Option[T]. That should allow us at compile time to verify the property you are checking at runtime below.

File: scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala [hunk 2]
```diff
@@ -91,9 +91,9 @@ case class Sketched[K, V](pipe: TypedPipe[(K, V)],
 case class SketchJoined[K: Ordering, V, V2, R](left: Sketched[K, V],
   right: TypedPipe[(K, V2)],
   numReducers: Int)(joiner: (K, V, Iterable[V2]) => Iterator[R])
-  extends HasReducers {
+  extends MustHasReducers {
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: this is both, right? HasReducers and MustHasReducers? Or MustHasReducers should extend HasReducers.

Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: I actually prefer the latter (MustHas extends Has).

Comment (reconditesea) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/Sketched.scala]: I meant to make MustHas extends Has, but forgot to actually do it in the code.
Will update this.

File: scalding-core/src/main/scala/com/twitter/scalding/typed/WithReducers.scala [hunk 3]
```diff
@@ -24,6 +24,14 @@ trait HasReducers {
 }
 
 /**
+ * used for types that must know how many reducers they need
+ * e.g. Sketched
+ */
+trait MustHasReducers {
```
Comment (azymnis) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/WithReducers.scala]: Sorry to be picky. Can we call this `MustHaveReducers` instead?

Comment (reconditesea) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/WithReducers.scala]: Of course!

File: scalding-core/src/main/scala/com/twitter/scalding/typed/WithReducers.scala [hunk 4]
```diff
@@ -24,6 +24,14 @@ trait HasReducers {
 }
 
 /**
+ * used for types that must know how many reducers they need
+ * e.g. Sketched
+ */
+trait MustHaveReducers[+This <: MustHaveReducers[This]] extends HasReducers {
```
Comment (johnynek) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/WithReducers.scala]: why do we need the F-bounded polymorphism here? The trait never references `This`. I think this complicates things a bit. Can it be removed?

**Pull request merged**